tests/infrastructure/prometheus.go

```diff
@@ -0,0 +1,688 @@
+		Expect(err).ToNot(HaveOccurred())
+		l, err := labels.Parse("prometheus.kubevirt.io=true")
+		Expect(err).ToNot(HaveOccurred())
+		pods, err := virtClient.CoreV1().Pods(flags.KubeVirtInstallNamespace).List(context.Background(), metav1.ListOptions{LabelSelector: l.String()})
+		Expect(err).ToNot(HaveOccurred())
+		Expect(endpoint.Subsets).To(HaveLen(1))
+
+		By("checking if the endpoint contains the metrics port and only one matching subset")
+		Expect(endpoint.Subsets[0].Ports).To(HaveLen(1))
+		Expect(endpoint.Subsets[0].Ports[0].Name).To(Equal("metrics"))
+		Expect(endpoint.Subsets[0].Ports[0].Port).To(Equal(int32(8443)))
+
+		By("checking if  the IPs in the subset match the KubeVirt system Pod count")
+		Expect(len(pods.Items)).To(BeNumerically(">=", 3), "At least one api, controller and handler need to be present")
+		Expect(endpoint.Subsets[0].Addresses).To(HaveLen(len(pods.Items)))
+
+		ips := map[string]string{}
+		for _, ep := range endpoint.Subsets[0].Addresses {
+			ips[ep.IP] = ""
+		}
+		for _, pod := range pods.Items {
+			Expect(ips).To(HaveKey(pod.Status.PodIP), fmt.Sprintf("IP of Pod %s not found in metrics endpoint", pod.Name))
+		}
+	})
+	It("[test_id:4139]should return Prometheus metrics", func() {
+		endpoint, err := virtClient.CoreV1().Endpoints(flags.KubeVirtInstallNamespace).Get(context.Background(), "kubevirt-prometheus-metrics", metav1.GetOptions{})
+		Expect(err).ToNot(HaveOccurred())
+		for _, ep := range endpoint.Subsets[0].Addresses {
+			cmd := fmt.Sprintf("curl -L -k https://%s:8443/metrics", tests.FormatIPForURL(ep.IP))
+			stdout, stderr, err := exec.ExecuteCommandOnPodWithResults(virtClient, pod, "virt-handler", strings.Fields(cmd))
+			Expect(err).ToNot(HaveOccurred(), fmt.Sprintf(remoteCmdErrPattern, cmd, stdout, stderr, err))
+			Expect(stdout).To(ContainSubstring("go_goroutines"))
+		}
+	})
+
+	DescribeTable("should throttle the Prometheus metrics access", func(family k8sv1.IPFamily) {
+		libnet.SkipWhenClusterNotSupportIPFamily(family)
+
+		ip := libinfra.GetSupportedIP(handlerMetricIPs, family)
+
+		if netutils.IsIPv6String(ip) {
+			Skip("Skip testing with IPv6 until https://github.com/kubevirt/kubevirt/issues/4145 is fixed")
+		}
+
+		concurrency := 100 // random value "much higher" than maxRequestsInFlight
+
+		tr := &http.Transport{
+			MaxIdleConnsPerHost: concurrency,
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
+			},
+		}
+
+		client := http.Client{
+			Timeout:   time.Duration(1 * time.Second),
+			Transport: tr,
+		}
+
+		errorsChan := make(chan error)
+		By("Scraping the Prometheus endpoint")
+		metricsURL := tests.PrepareMetricsURL(ip, 8443)
+		for ix := 0; ix < concurrency; ix++ {
+			go func(ix int) {
+				req, _ := http.NewRequest("GET", metricsURL, nil)
+				resp, err := client.Do(req)
+				if err != nil {
+					_, err = fmt.Fprintf(GinkgoWriter, "client: request: %v #%d: %v\n", req, ix, err) // troubleshooting helper
```

Comment:
I think you need to change the name, not to re-use the error variable. Just because the orig error is used for something else (send to a channel). By itself it is not wrong, because in case of an error here, the expect will stop everything.

Reply:
Great spot! The test failure actually revealed that I broke the test logic here. Fixed.